File: python/1269.number-of-ways-to-stay-in-the-same-place-after-some-steps.py
class Solution:
    def numWays(self, steps: int, arrLen: int) -> int:
        MOD = 10**9 + 7

        arrLen = min(steps + 1, arrLen)
        
        @cache
        def dp(steps: int, pos: int) -> int:
            if pos >= arrLen or pos < 0 or steps < 0:
                return 0

            if pos == 0 and steps == 0:
                return 1

            res = dp(steps - 1, pos + 1) + dp(steps - 1, pos - 1) + dp(steps - 1, pos)
            return res % MOD


        return dp(steps, 0)

File: cpp/1269.number-of-ways-to-stay-in-the-same-place-after-some-steps.cpp
class Solution {
public:
    int numWays(int steps, int arrLen) {
        int MOD = 1e9 + 7;
        arrLen = min(steps / 2 + 1, arrLen);
        vector<vector<int> > dp(steps + 1, vector<int>(arrLen, 0));

        // base case
        dp[0][0] = 1;

        for (int step = 1; step <= steps; step++) {
            for (int pos = arrLen - 1; pos >= 0; pos--) {
                long res = dp[step - 1][pos];
                if (pos > 0) {
                    res += dp[step - 1][pos - 1];
                }
                if (pos < arrLen - 1) {
                    res += dp[step - 1][pos + 1];
                }
                dp[step][pos] = res % MOD;
            }
        }
        return dp[steps][0];
    }
};